File: conformance/tests/ec-demo/ec-demo.runar.zig
const runar = @import("runar");

pub const ECDemo = struct {
    pub const Contract = runar.SmartContract;

    pt: runar.Point,

    pub fn init(pt: runar.Point) ECDemo {
        return .{ .pt = pt };
    }

    pub fn checkX(self: *const ECDemo, expectedX: i64) void {
        runar.assert(runar.ecPointX(self.pt) == expectedX);
    }

    pub fn checkY(self: *const ECDemo, expectedY: i64) void {
        runar.assert(runar.ecPointY(self.pt) == expectedY);
    }

    pub fn checkMakePoint(self: *const ECDemo, x: i64, y: i64, expectedX: i64, expectedY: i64) void {
        _ = self;
        const p = runar.ecMakePoint(x, y);
        runar.assert(runar.ecPointX(p) == expectedX);
        runar.assert(runar.ecPointY(p) == expectedY);
    }

    pub fn checkOnCurve(self: *const ECDemo) void {
        runar.assert(runar.ecOnCurve(self.pt));
    }

    pub fn checkAdd(self: *const ECDemo, other: runar.Point, expectedX: i64, expectedY: i64) void {
        const result = runar.ecAdd(self.pt, other);
        runar.assert(runar.ecPointX(result) == expectedX);
        runar.assert(runar.ecPointY(result) == expectedY);
    }

    pub fn checkMul(self: *const ECDemo, scalar: i64, expectedX: i64, expectedY: i64) void {
        const result = runar.ecMul(self.pt, scalar);
        runar.assert(runar.ecPointX(result) == expectedX);
        runar.assert(runar.ecPointY(result) == expectedY);
    }

    pub fn checkMulGen(self: *const ECDemo, scalar: i64, expectedX: i64, expectedY: i64) void {
        _ = self;
        const result = runar.ecMulGen(scalar);
        runar.assert(runar.ecPointX(result) == expectedX);
        runar.assert(runar.ecPointY(result) == expectedY);
    }

    pub fn checkNegate(self: *const ECDemo, expectedNegY: i64) void {
        const neg = runar.ecNegate(self.pt);
        runar.assert(runar.ecPointY(neg) == expectedNegY);
    }

    pub fn checkNegateRoundtrip(self: *const ECDemo) void {
        const neg1 = runar.ecNegate(self.pt);
        const neg2 = runar.ecNegate(neg1);
        runar.assert(runar.ecPointX(neg2) == runar.ecPointX(self.pt));
        runar.assert(runar.ecPointY(neg2) == runar.ecPointY(self.pt));
    }

    pub fn checkModReduce(self: *const ECDemo, value: i64, modulus: i64, expected: i64) void {
        _ = self;
        runar.assert(runar.ecModReduce(value, modulus) == expected);
    }

    pub fn checkEncodeCompressed(self: *const ECDemo, expected: runar.ByteString) void {
        const compressed = runar.ecEncodeCompressed(self.pt);
        runar.assert(runar.bytesEq(compressed, expected));
    }

    pub fn checkMulIdentity(self: *const ECDemo) void {
        const result = runar.ecMul(self.pt, 1);
        runar.assert(runar.ecPointX(result) == runar.ecPointX(self.pt));
        runar.assert(runar.ecPointY(result) == runar.ecPointY(self.pt));
    }

    pub fn checkAddOnCurve(self: *const ECDemo, other: runar.Point) void {
        const result = runar.ecAdd(self.pt, other);
        runar.assert(runar.ecOnCurve(result));
    }

    pub fn checkMulGenOnCurve(self: *const ECDemo, scalar: i64) void {
        _ = self;
        const result = runar.ecMulGen(scalar);
        runar.assert(runar.ecOnCurve(result));
    }
};

File: conformance/tests/ec-demo/ec-demo.runar.go
//go:build ignore

package contract

import "runar"

type ECDemo struct {
	runar.SmartContract
	Pt runar.Point `runar:"readonly"`
}

func (c *ECDemo) CheckX(expectedX runar.Bigint) {
	runar.Assert(runar.EcPointX(c.Pt) == expectedX)
}

func (c *ECDemo) CheckY(expectedY runar.Bigint) {
	runar.Assert(runar.EcPointY(c.Pt) == expectedY)
}

func (c *ECDemo) CheckMakePoint(x, y, expectedX, expectedY runar.Bigint) {
	p := runar.EcMakePoint(x, y)
	runar.Assert(runar.EcPointX(p) == expectedX)
	runar.Assert(runar.EcPointY(p) == expectedY)
}

func (c *ECDemo) CheckOnCurve() {
	runar.Assert(runar.EcOnCurve(c.Pt))
}

func (c *ECDemo) CheckAdd(other runar.Point, expectedX, expectedY runar.Bigint) {
	result := runar.EcAdd(c.Pt, other)
	runar.Assert(runar.EcPointX(result) == expectedX)
	runar.Assert(runar.EcPointY(result) == expectedY)
}

func (c *ECDemo) CheckMul(scalar, expectedX, expectedY runar.Bigint) {
	result := runar.EcMul(c.Pt, scalar)
	runar.Assert(runar.EcPointX(result) == expectedX)
	runar.Assert(runar.EcPointY(result) == expectedY)
}

func (c *ECDemo) CheckMulGen(scalar, expectedX, expectedY runar.Bigint) {
	result := runar.EcMulGen(scalar)
	runar.Assert(runar.EcPointX(result) == expectedX)
	runar.Assert(runar.EcPointY(result) == expectedY)
}

func (c *ECDemo) CheckNegate(expectedNegY runar.Bigint) {
	neg := runar.EcNegate(c.Pt)
	runar.Assert(runar.EcPointY(neg) == expectedNegY)
}

func (c *ECDemo) CheckNegateRoundtrip() {
	neg1 := runar.EcNegate(c.Pt)
	neg2 := runar.EcNegate(neg1)
	runar.Assert(runar.EcPointX(neg2) == runar.EcPointX(c.Pt))
	runar.Assert(runar.EcPointY(neg2) == runar.EcPointY(c.Pt))
}

func (c *ECDemo) CheckModReduce(value, modulus, expected runar.Bigint) {
	runar.Assert(runar.EcModReduce(value, modulus) == expected)
}

func (c *ECDemo) CheckEncodeCompressed(expected runar.ByteString) {
	compressed := runar.EcEncodeCompressed(c.Pt)
	runar.Assert(compressed == expected)
}

func (c *ECDemo) CheckMulIdentity() {
	result := runar.EcMul(c.Pt, 1)
	runar.Assert(runar.EcPointX(result) == runar.EcPointX(c.Pt))
	runar.Assert(runar.EcPointY(result) == runar.EcPointY(c.Pt))
}

func (c *ECDemo) CheckAddOnCurve(other runar.Point) {
	result := runar.EcAdd(c.Pt, other)
	runar.Assert(runar.EcOnCurve(result))
}

func (c *ECDemo) CheckMulGenOnCurve(scalar runar.Bigint) {
	result := runar.EcMulGen(scalar)
	runar.Assert(runar.EcOnCurve(result))
}
